docs/topics/testing/tools.txt

```diff
@@ -880,6 +880,32 @@ It also provides an additional method:
     modify them, you could reload them in the ``setUp()`` method with
     :meth:`~django.db.models.Model.refresh_from_db`, for example.
 
+.. note::
+
+    Testing code which uses :func:`~django.db.transaction.on_commit` hooks are
+    not run during the execution of :class:`~django.test.TestCase` tests. As
+    mentioned above, the proper way to test this is to use the
+    :class:`~django.test.TransactionTestCase`. However, should it be required
+    to trigger :func:`~django.db.transaction.on_commit` in a
+    :class:`~django.test.TestCase` test, one possible solution is to add
+    a context manager as follows::
+
+        @contextmanager
+        def immediate_on_commit(self, using=None):
+            """
+            Context manager executing transaction.on_commit() hooks immediately as
+            if the connection was in auto-commit mode. This is required when
+            using a subclass of django.test.TestCase as all tests are wrapped in
+            a transaction that never gets committed.
+            """
+            immediate_using = DEFAULT_DB_ALIAS if using is None else using
+            def on_commit(func, using=None):
+                using = DEFAULT_DB_ALIAS if using is None else using
+                if using == immediate_using:
+                    func()
+            with mock.patch('django.db.transaction.on_commit', side_effect=on_commit) as patch:
+                yield patch
```

Comment:
fwiw this approach won't catch the case described in https://code.djangoproject.com/ticket/30457#comment:10 hence why I think it's more appropriate to run callbacks when yielded back execution instead of immediately by default.